### PR TITLE
docs: `when-then-otherwise` examples

### PR DESCRIPTION
**Pull request by dangotbanned**
Follow-up referenced in https://github.com/vega/altair/pull/3427#issuecomment-2237616441

Developing each function/method doc further with example(s)

# Tasks
- [x] `empty` note (adapted from https://github.com/vega/altair/pull/3490)
- [x] Each method has a situational example
	- [x] `Then.when`
	- [x] `Then.otherwise`
	- [x] `When.then`
	- [x] `ChainedWhen.then`
- [x] Multiple examples
	- [x] `alt.when` (using `vega_datasets`)
- [x] Adding a `__repr__` for `When`, `ChainedWhen`
	- `Then` inherits one

## Possible tasks
Just some ideas that came up while working on this. 
Can work on them if there's interest, but not planning to otherwise:
- Moving parts of the **Parameters** section into examples
	- E.g. `empty` note

# Comments from original PR
Linking these as I raised them as potential issues, but they were not discussed:
- [`Then` method autocomplete visibility](https://github.com/vega/altair/pull/3427#discussion_r1668638763)
- [[`When | Then | ChainedWhen`] method docs not in API reference](https://github.com/vega/altair/pull/3427#discussion_r1659965947)
- [How should `polars` be referenced and/or accredited for idea?](https://github.com/vega/altair/pull/3427#discussion_r1662514935)

### CODE REVIEW

**Comment by mattijn** [comment]
Thanks! This looks great. I was expecting these examples to be around here: https://altair-viz.github.io/user_guide/interactions.html#conditions-filters, but the examples in this PR are part of the docstrings, which is very nice too! We should have more of these.

Will add a note to https://github.com/vega/altair/issues/3500.

I noticed when playing with this that I agree with your suggestion that it would be nice to have a `__repr__` for `When` and `ChainedWhen`.

Regarding, your three linked comments from original PR. I don't have clear insight on these either. I don't think it is necessary to use intersphinx, but a reference to the polars docs page of the `pl.when` syntax makes sense though.

**Comment by dangotbanned** [comment]
> Thanks! This looks great. I was expecting these examples to be around here: [altair-viz.github.io/user_guide/interactions.html#conditions-filters](https://altair-viz.github.io/user_guide/interactions.html#conditions-filters), but the examples in this PR are part of the docstrings, which is very nice too! We should have more of these.
> 
> Will add a note to #3500.
> 

Thank you for reviewing, but my bad for misunderstanding the assignment @mattijn

Hopefully they will make writing the *User Guide* parts easier in #3500, since most of the examples already produce a valid chart.
The [last one](https://github.com/dangotbanned/altair/blob/0e7b49f45807aac494861465b9a734ea8e74def0/altair/vegalite/v5/api.py#L1193) for `alt.when` I wanted to adapt - but the `cars` dataset wasn't a good fit.

---

> I noticed when playing with this that I agree with your suggestion that it would be nice to have a `__repr__` for `When` and `ChainedWhen`.

For reference, following [`0e7b49f` (#3492)](https://github.com/vega/altair/pull/3492/commits/0e7b49f45807aac494861465b9a734ea8e74def0)

image

The `ChainedWhen` one is tricky as it represents 
- a list of complete conditions (`self._conditions`)
- a single partial condition (`self._condition`), prior to the next `.then()`

**Comment by dangotbanned** [comment]
@mattijn are we good to go with this one?

Sorry I can't seem to request a review

File: altair/vegalite/v5/api.py [hunk 1]
```diff
@@ -1066,9 +1157,40 @@ def when(
...
+            alt.Chart(source)
+            .mark_point()
+            .encode(x="Horsepower", y="Miles_per_Gallon")
+            .add_params(brush)
+        )
+        points
+
+    Basic ``if-then-else`` conditions translate directly to ``when-then-otherwise``::
+
+        points.encode(color=alt.when(brush).then("Origin").otherwise(alt.value("lightgray"))
```
Comment (mattijn) [inline altair/vegalite/v5/api.py:1177]:
```suggestion
        points.encode(color=alt.when(brush).then("Origin").otherwise(alt.value("lightgray")))
```

**Review by mattijn — APPROVED** [review]
Looks great. One suggested change. 

I like the examples. Really clean.

```python
# Simple conditions may be expressed without defining a default
import altair as alt
from vega_datasets import data

source = data.movies()
predicate = (alt.datum.IMDB_Rating == None) | (alt.datum.Rotten_Tomatoes_Rating == None)

alt.Chart(source).mark_point(invalid=None).encode(
    x="IMDB_Rating:Q",
    y="Rotten_Tomatoes_Rating:Q",
    color=alt.when(predicate).then(alt.value("grey")),
)
```



```python
# Points outside of `brush` will not appear highlighted
import altair as alt
from vega_datasets import data

source = data.cars()
brush = alt.selection_interval()
color = alt.when(brush).then("Origin:N").otherwise(alt.value("grey"))

alt.Chart(source).mark_point().encode(
    x="Horsepower:Q",
    y="Miles_per_Gallon:Q",
    color=color,
).add_params(brush)
```


```python
 # Chain calls to express precise queries
import altair as alt
from vega_datasets import data

source = data.cars()
color = (
    alt.when(alt.datum.Miles_per_Gallon >= 30, Origin="Europe")
    .then(alt.value("crimson"))
    .when(alt.datum.Horsepower > 150)
    .then(alt.value("goldenrod"))
    .otherwise(alt.value("grey"))
)

alt.Chart(source).mark_point().encode(x="Horsepower", y="Miles_per_Gallon", color=color)
```


```python
# Multiple conditions with an implicit default
import altair as alt
from vega_datasets import data

source = data.movies()
predicate = (alt.datum.IMDB_Rating == None) | (alt.datum.Rotten_Tomatoes_Rating == None)
color = (
    alt.when(predicate)
    .then(alt.value("grey"))
    .when(alt.datum.IMDB_Votes < 5000)
    .then(alt.value("lightblue"))
)

alt.Chart(source).mark_point(invalid=None).encode(
    x="IMDB_Rating:Q", y="Rotten_Tomatoes_Rating:Q", color=color
)
```


```python
# Setting up a common chart
import altair as alt
from vega_datasets import data

source = data.cars()
brush = alt.selection_interval()

points = (
    alt.Chart(source)
    .mark_point()
    .encode(x="Horsepower", y="Miles_per_Gallon")
    .add_params(brush)
)
points
```


```python
# Basic `if-then-else` conditions translate directly to `when-then-otherwise`
points.encode(color=alt.when(brush).then("Origin").otherwise(alt.value("lightgray")))
```


```python
# Omitting the `.otherwise()` clause will use the channel default instead
points.encode(color=alt.when(brush).then("Origin"))
```


```python
# Predicates passed as positional arguments will be reduced with `&`
points.encode(
    color=alt.when(
        brush, (alt.datum.Miles_per_Gallon >= 30) | (alt.datum.Horsepower >= 130)
    )
    .then("Origin")
    .otherwise(alt.value("lightgray"))
)
```


```python
# Using keyword-argument `constraints` can simplify compositions like
verbose_composition = (
    (alt.datum.Name == "Name_1")
    & (alt.datum.Color == "Green")
    & (alt.datum.Age == 25)
    & (alt.datum.StartDate == "2000-10-01")
)
when_verbose = alt.when(verbose_composition)
# To
when_concise = alt.when(Name="Name_1", Color="Green", Age=25, StartDate="2000-10-01")
```

**Comment by dangotbanned** [comment]
Thanks @mattijn for the review and well spotted on the missing `)`

Seeing them all together in https://github.com/vega/altair/pull/3492#pullrequestreview-2205990789 really looks like this will fit into the User Guide nicely

**Comment by mattijn** [comment]
Yes! Beautiful new addition to the Altair library! Well done @dangotbanned!

**Pull request merged**